File: ru.kmorozov.library.data.model/src/test/java/ru/kmorozov/library/data/test/mongo/BaseOperationsTest.java
package ru.kmorozov.library.data.test.mongo;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.TextCriteria;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import ru.kmorozov.library.data.config.MongoConfiguration;
import ru.kmorozov.library.data.model.book.Book;
import ru.kmorozov.library.data.repository.BooksRepository;
import ru.kmorozov.library.data.storage.mongo.LikeTextSearch;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import static org.hamcrest.CoreMatchers.*;

/**
 * Created by sbt-morozov-kv on 28.11.2016.
 */

@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(classes = MongoConfiguration.class)
public class BaseOperationsTest {

    @Autowired
    BooksRepository booksRepository;

    @Autowired
    private MongoTemplate mongoTemplate;

    @Test
    public void connectTest() {
        Assert.assertThat(this.booksRepository, notNullValue());
        this.booksRepository.deleteAll();
        Assert.assertThat(this.booksRepository.count(), is(0));
    }

    @Test
    public void crudTest() {
        long countBefore = this.booksRepository.count();

        Book book = new Book("Test title", "Test author");

        Book savedBook = this.booksRepository.save(book);
        Assert.assertThat(savedBook, notNullValue());
        Assert.assertThat(savedBook, is(book));
        Assert.assertThat(this.booksRepository.count(), is(countBefore + 1L));
        this.booksRepository.delete(book);
        Assert.assertThat(this.booksRepository.count(), is(countBefore));
    }

    @Test
    public void searchTest() {
        List<Book> books = Arrays.asList(new Book("Test tit1le", "Test aut1hor"), new Book("Test tit2le", "Test aut2hor"));

        this.booksRepository.saveAll(books);

        try {
            LikeTextSearch likeTextSearch = new LikeTextSearch(Book.class.getSimpleName(), this.mongoTemplate);
            Assert.assertThat(likeTextSearch.findMatchingIds("%aut1%").size(), is(0));
            TextCriteria criteria2 = TextCriteria.forDefaultLanguage().matching("Test");
            Assert.assertThat(this.booksRepository.findAllBy(criteria2).size(), is(2));
        } finally {
            this.booksRepository.deleteAll(books);
        }
    }

    @Test
    public void findLinks() {
        Stream<Book> lnkBooks = this.booksRepository.streamByBookInfoFormat("LNK");
        Assert.assertTrue(10L > lnkBooks.count());
    }
}

File: ru.kmorozov.library.data.model/src/test/java/ru/kmorozov/library/data/test/mongo/BaseOperationsTest.kt
package ru.kmorozov.library.data.test.mongo

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.query.TextCriteria
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner
import ru.kmorozov.library.data.config.MongoConfiguration
import ru.kmorozov.library.data.model.book.Book
import ru.kmorozov.library.data.model.book.BookInfo
import ru.kmorozov.library.data.repository.BooksRepository
import ru.kmorozov.library.data.storage.mongo.LikeTextSearch
import java.util.*

/**
 * Created by sbt-morozov-kv on 28.11.2016.
 */

@RunWith(SpringJUnit4ClassRunner::class)
@ContextConfiguration(classes = arrayOf(MongoConfiguration::class))
class BaseOperationsTest {

    @Autowired
    internal lateinit var booksRepository: BooksRepository

    @Autowired
    private lateinit var mongoTemplate: MongoTemplate

    @Test
    fun connectTest() {
        MatcherAssert.assertThat<BooksRepository>(booksRepository, notNullValue())
        booksRepository.deleteAll()
        MatcherAssert.assertThat(booksRepository.count(), `is`<Number>(0))
    }

    @Test
    fun crudTest() {
        val countBefore = booksRepository.count()

        val book = Book("Test title", "Test author")

        val savedBook = booksRepository.save(book)
        MatcherAssert.assertThat(savedBook, notNullValue())
        MatcherAssert.assertThat(savedBook, `is`(book))
        MatcherAssert.assertThat(booksRepository.count(), `is`(countBefore + 1L))
        booksRepository.delete(book)
        MatcherAssert.assertThat(booksRepository.count(), `is`(countBefore))
    }

    @Test
    fun searchTest() {
        val books = Arrays.asList(Book("Test tit1le", "Test aut1hor"), Book("Test tit2le", "Test aut2hor"))

        booksRepository.saveAll(books)

        try {
            val likeTextSearch = LikeTextSearch(Book::class.java.simpleName, mongoTemplate)
            MatcherAssert.assertThat(likeTextSearch.findMatchingIds("%aut1%").size, `is`(0))
            val criteria2 = TextCriteria.forDefaultLanguage().matching("Test")
            MatcherAssert.assertThat(booksRepository.findAllBy(criteria2).size, `is`(2))
        } finally {
            booksRepository.deleteAll(books)
        }
    }

    @Test
    fun findLinks() {
        val lnkBooks = booksRepository.streamByBookInfoFormat(BookInfo.BookFormat.LNK)
        Assert.assertTrue(10L > lnkBooks.count())
    }
}
